File: verifier-js/lib/did.js
'use strict';

// Minimal `did:key` resolution for Ed25519 keys (W3C did:key method).
// A did:key is self-contained — the public key *is* the identifier:
//
//   did:key:z<base58btc( 0xED 0x01 || <32-byte ed25519 pubkey> )>
//
// so resolving one needs no network and no registry.

const B58_ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

const B58_INDEX = new Map();
for (let i = 0; i < B58_ALPHABET.length; i++) B58_INDEX.set(B58_ALPHABET[i], i);

function base58Decode(str) {
  let leadingOnes = 0;
  while (leadingOnes < str.length && str[leadingOnes] === '1') leadingOnes++;

  let num = 0n;
  for (const ch of str) {
    const v = B58_INDEX.get(ch);
    if (v === undefined) throw new Error(`invalid base58 character: ${JSON.stringify(ch)}`);
    num = num * 58n + BigInt(v);
  }

  const body = [];
  while (num > 0n) {
    body.unshift(Number(num & 0xffn));
    num >>= 8n;
  }

  return Buffer.concat([Buffer.alloc(leadingOnes, 0), Buffer.from(body)]);
}

const DID_KEY_PREFIX = 'did:key:z';
const ED25519_MULTICODEC = Buffer.from([0xed, 0x01]);

// Resolve an Ed25519 did:key to its 32-byte raw public key (Buffer).
// Throws on anything that is not a well-formed Ed25519 did:key.
function resolveEd25519Did(did) {
  if (typeof did !== 'string' || !did.startsWith(DID_KEY_PREFIX)) {
    throw new Error('not an Ed25519 did:key');
  }

  const decoded = base58Decode(did.slice(DID_KEY_PREFIX.length));
  if (decoded.length !== 34 || !decoded.subarray(0, 2).equals(ED25519_MULTICODEC)) {
    throw new Error('unsupported did:key multicodec (expected ed25519-pub)');
  }

  return decoded.subarray(2);
}

module.exports = { base58Decode, resolveEd25519Did, DID_KEY_PREFIX };

File: verifier-rs/src/did.rs
//! Minimal `did:key` resolution for Ed25519 keys.
//!
//! A did:key is self-contained — the public key *is* the identifier:
//!
//! ```text
//! did:key:z<base58btc( 0xED 0x01 || <32-byte ed25519 pubkey> )>
//! ```

const ALPHABET: &[u8; 58] = b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

/// Decode a base58btc (Bitcoin alphabet) string. Returns `None` on any
/// character outside the alphabet.
pub fn base58_decode(input: &str) -> Option<Vec<u8>> {
    let mut bytes: Vec<u8> = Vec::new();

    for ch in input.bytes() {
        let mut carry = ALPHABET.iter().position(|&a| a == ch)? as u32;
        for byte in bytes.iter_mut().rev() {
            carry += (*byte as u32) * 58;
            *byte = (carry & 0xff) as u8;
            carry >>= 8;
        }
        while carry > 0 {
            bytes.insert(0, (carry & 0xff) as u8);
            carry >>= 8;
        }
    }

    // leading '1's encode leading zero bytes
    for ch in input.bytes() {
        if ch == b'1' {
            bytes.insert(0, 0);
        } else {
            break;
        }
    }

    Some(bytes)
}

const DID_KEY_PREFIX: &str = "did:key:z";
const ED25519_MULTICODEC: [u8; 2] = [0xed, 0x01];

/// Resolve an Ed25519 `did:key` to its 32-byte raw public key.
pub fn resolve_ed25519_did(did: &str) -> Option<[u8; 32]> {
    let b58 = did.strip_prefix(DID_KEY_PREFIX)?;
    let decoded = base58_decode(b58)?;
    if decoded.len() != 34 || decoded[0..2] != ED25519_MULTICODEC {
        return None;
    }
    let mut key = [0u8; 32];
    key.copy_from_slice(&decoded[2..34]);
    Some(key)
}
